lib/connections/service.go

```diff
@@ -490,15 +490,43 @@ func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Con
 	// Perform dials according to the queue, stopping when we've reached the
 	// allowed additional number of connections (if limited).
 	numConns := 0
-	for _, entry := range queue {
-		if conn, ok := s.dialParallel(ctx, entry.id, entry.targets); ok {
-			s.conns <- conn
-			numConns++
-			if allowAdditional > 0 && numConns >= allowAdditional {
+	numDials := 0
+	dialCond := stdsync.NewCond(new(stdsync.Mutex))
+	dialWG := new(stdsync.WaitGroup)
+	for i := range queue {
+		select {
+		case <-ctx.Done():
+			break
+		default:
+		}
+		if allowAdditional > 0 {
+			done := false
+			dialCond.L.Lock()
+			for numDials+numConns >= allowAdditional && !done {
+				dialCond.Wait()
+				done = numConns >= allowAdditional
+			}
+			dialCond.L.Unlock()
+			if done {
 				break
 			}
 		}
+		numDials++
+		dialWG.Add(1)
+		go func(entry dialQueueEntry) {
```

Comment:
I think we should have a semaphore here with a reasonable limit (32?) so that we don't cause issues or explode our FD usage when dialling thousands of devices

Reply:
Sounds reasonable. Given the FDs apply per dial, not device, I moved the semaphore/limitation there. And to prevent the case that one device has so many addresses to dial that it "clogs up" the semaphore, there's a two step limitation: 8 per device, 64 globally. I reused `model.byteSemaphore` for that which is now `util.Semaphore`, and introduced `util.MultiSemaphore` which combines multiple semaphores the same way `model.newLimitedRequestResponse` already did. That blew up the diff, but really it's just renaming.